### PR TITLE
Enforce lifetime storage across retained document versions

### PR DESCRIPTION
**Pull request by logancyang**
Fixes Brevilabs/app-sites#555

## Why

Limiting live documents does not limit their retained HTML history. A Publisher with an explicit lifetime allowance needs uploads to stop at 5 GiB without taking existing pages offline or narrowing existing customers' rights.

## What

| Situation | Result |
| --- | --- |
| Explicit lifetime allowance | 5 GiB across retained versions and outstanding uploads; existing document, daily and per-upload ceilings remain |
| Concurrent or linked-identity uploads | Share one byte allowance, reserved before storage |
| Allowance full | Further uploads get cleanup guidance; pages, history, listing and withdrawal still work |
| Successful withdrawal | Confirmed removed bytes become available again |
| Uncertain write or cleanup | Bytes remain charged until reconciliation, rather than silently escaping the allowance |
| Existing plan without the explicit allowance | Existing limits remain unchanged |

Account output includes used bytes and, when applicable, allowed bytes. Older clients continue to work. Rejected creates cannot erase another upload or a withdrawal racing their rollback.

## Non goal

No automatic deletion, retention scheduler, deduplication, forced subscription, overage billing, new dashboard, or retroactive limits on free/monthly/Copilot/manual/self-host plans. No payment integration, hosted prices, payment identifiers or private economics. Deployment, production migration execution, sales activation and CLI publication are separate operations.

## Screenshot

Not applicable: this changes upload enforcement and machine-readable account output, not a rendered interface.

## Risk

**High**

| Criterion | Status | Reason |
| --- | :---: | --- |
| No behavior change, or a cosmetic/copy/docs/config change visible where it renders, or deterministic tests cover the changed behavior | ✅ | Real D1/R2 tests cover cap boundaries, concurrency, migration, failures and cleanup |
| A defect would fail CI or be obvious on first use | ❌ | Historical orphan completeness also requires an authorized R2 inventory check |
| A revert fully restores prior state, including persisted data | ❌ | Adds persisted byte reservations; rollback must preserve existing capped grants |
| No auth, permissions, secrets, or input-handling surface changes | ❌ | Trusted entitlement selects a distinct capped plan |
| No public API, plugin API, message, or on-disk contract changes | ❌ | Adds optional allowance and account fields plus a D1 table |
| No core-path concurrency, async-lifecycle, or state-machine changes | ❌ | Capacity reservations span upload and withdrawal outcomes |
| No hot-path behavior lacks deterministic coverage | ✅ | Upload, rejection, linked-owner and rollback interleavings are tested |
| No new dependency | ✅ | Uses existing D1/R2 bindings and test runtime |
| Human-only behavior stays in one feature area and surfaces quickly | ❌ | Historical accounting discrepancies can remain hidden until capacity is reached |

Review: inspect `reserveStorage`/`releaseStorage` in `src/storage-quota.ts`, `storeVersion` in `src/api/push.ts`, `rollbackCreate` in `src/db.ts`, and the activation/recovery requirements in `docs/storage-allowance.md`.

## Verification

1. Run `npm run typecheck`, `node scripts/check-account-actions.cjs`, `node --test test/release-workflow.node.cjs`, and `npm test` in this branch.
2. Run `npx vitest run test/storage-quota.test.ts`: its bounded local fixture fills the allowance using multiple UTF-8 versions, rejects one extra byte, reads current/history pages, withdraws the document and successfully uploads again.
3. Inspect the same suite's concurrent upload, linked identity, failed put/metadata/delete, historical backfill and rejected-create rollback cases; inspect the entitlement and CLI tests for existing-plan and older-response compatibility.
4. Before enabling the allowance, follow the authorized deployment and full historical inventory check in `docs/storage-allowance.md`. The migration alone is not proof that every historical R2 object is accounted for.

### CODE REVIEW

**Comment by logancyang** [comment]
@codex review

Please review the current head, focusing on retained-byte accounting across R2 writes, D1 metadata, withdrawal/concurrent-upload interleavings, and preservation of existing customer entitlements.

File: src/api/push.ts [hunk 1]
```diff
@@ -165,14 +165,20 @@ async function storeVersion(
...
+  limits: PlanLimits | null,
+): Promise<"stored" | "deleted" | "full"> {
   // The publisher's own bytes, unmodified. OpenArtifacts' additions go in when the
   // document is served, so a byline change — or a plan that removes one —
   // reaches documents already published. `size` is therefore the size of what
   // the publisher sent, not of what a reader receives.
   const bytes = new TextEncoder().encode(html);
   const key = versionObjectKey(docId, version);
 
+  if (!(await reserveStorage(env.DB, publisher.owner, docId, version, bytes.byteLength, limits?.storageBytes))) {
```
[the remark this answers is not in the document]
Comment (logancyang) [reply src/api/push.ts:178]: Not a defect under this PR’s existing-customer contract. A legacy key cannot publish merely because its owner bought lifetime: `authenticateRequest` validates that separate credential, and `handleApi` requires an independently eligible `plus`/`believer` Copilot plan before create/update. Applying the linked account’s lifetime cap to that independently valid grant would violate the explicit non-goal of retroactively restricting Copilot/legacy-license rights.

I reproduced the reported over-cap behavior with real Miniflare D1/R2 and verified the intended distinction: linked Plus/Believer keys can create/update under their independent uncapped grant, every byte remains in the same linked-owner ledger, the capped account credential is refused at capacity, and changing the legacy key to an ineligible plan rejects both create and update without adding bytes. A lifetime-only purchase does not create or authorize that legacy credential.

The billing slice must prefer independently valid uncapped access when resolving the combined account entitlement; that cross-service check remains part of the billing review. No Worker change to indiscriminately cap legacy credentials is warranted.

**Comment by logancyang** [comment]
No blocking code findings at 3ad9c716a55c9882429bc0d8dbd886ec59bd1fa5.

Verified the atomic retained-byte reservations, conservative failure accounting, exact-key deletion credits, linked-owner scope, additive account/CLI behavior, and preservation of independently valid uncapped rights. The Codex legacy-key finding is dispositioned inline and resolved; applying the lifetime cap to a separately eligible Copilot grant would violate this PR’s contract.

Checks: typechecks and account-actions check pass; Worker511/511 with bounded test concurrency, CLI17/17, release-workflow11/11, and three independent real-D1/R2 checks pass. The unrestricted local full run hit15 unchanged approval-fixture expiry failures; that suite passes in isolation and the complete head passes with bounded concurrency. CI is green. Latest-head Codex completed, with no unresolved threads or pending reviews.

Code budget: +440/-30 across16 files. The small reservation ledger is justified because version metadata alone cannot account for uncertain R2 writes; existing identity/plan/withdrawal paths are reused.

Operational sequencing remains separate: apply the additive migration before Worker deployment, and reconcile the full historical R2 inventory before enabling capped lifetime entitlements. This review did not execute or verify those production steps. Combined-source billing precedence will be checked in the billing PR.

**Comment by logancyang** [comment]
No blocking findings in conflict-resolved commit da520130a96d43a48584fcc245915c05b49b9bf6. Inspected both manual resolutions: hosted Free stays at three documents, the lifetime plan retains its 5 GiB allowance, and account output retains storedBytes. Storage enforcement, entitlement, migration, and CLI code remain unchanged from previously reviewed 3ad9c71. Independently passed all 513 Worker tests, 17 CLI tests, both typechecks, account-action validation, and release-workflow tests. This is code/integration approval, not verification of production R2 reconciliation or billing rollout.
